### PR TITLE
os/mac/diagnostic: fix OpenCore check

### LINKED ISSUE #18763: check_for_opencore false positive
### `brew doctor` output

```shell
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have installed macOS using OpenCore Legacy Patcher.
We do not provide support for this configuration.
It is expected behaviour that some formulae will fail to build in this unsupported configuration.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.4.5-14-gb421c2f
ORIGIN: https://github.com/Homebrew/brew
HEAD: b421c2f0685bbeab84aaa7f2f77a098bc8f68553
Last commit: 88 minutes ago
Core tap JSON: 13 Nov 10:13 UTC
Core cask tap JSON: 13 Nov 10:13 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: /usr/bin/nano
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_NO_ANALYTICS: set
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.6 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.47.0 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.1-arm64
CLT: 16.1.0.0.1.1729049160
Xcode: 16.1
Rosetta 2: false
```

### What were you trying to do (and why)?

[Newly added](https://github.com/Homebrew/brew/pull/18761) check_for_opencore check is falsely reporting that i've installed macOS using OpenCore Legacy Patcher. 
I have not, but I do have OpenCore Legacy patcher installed for purpose of making bootable usb. 

[The check](https://github.com/Homebrew/brew/blob/b421c2f0685bbeab84aaa7f2f77a098bc8f68553/Library/Homebrew/extend/os/mac/diagnostic.rb#L131) is wrong IMHO - it just checks for existance of those two files:

- `/Library/Application Support/Dortania/OpenCore-Patcher.app`
- `/Library/PrivilegedHelperTools/com.dortania.opencore-legacy-patcher.privileged-helper`

### What happened (include all command output)?

The warning showed up.

### What did you expect to happen?

No warning about OpenCore Legacy patcher.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew doctor
```

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Fixes #18763.

### CODE REVIEW

**Comment by cho-m** [comment]
Interesting that CI runner triggers it. I wonder if it is actually set up with OpenCore.

**Comment by ZhongRuoyu** [comment]
Yea, I only checked GitHub hosted runners which do not seem to use OpenCore.

If that's the case then maybe we can allow it in CI?

**Comment by cho-m** [comment]
Can you add `opencore_version` to message? Curious what it is outputting.

**Comment by ZhongRuoyu** [comment]
`Warning: You have installed macOS using OpenCore Legacy Patcher, version 4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102:opencore-version	REL-083-2022-08-01`

File: Library/Homebrew/extend/os/mac/diagnostic.rb [hunk 1]
```diff
@@ -129,11 +129,17 @@ def check_for_unsupported_macos
...
-          has_opencore ||= File.exist?("/Library/Application Support/Dortania/OpenCore-Patcher.app")
-          has_opencore ||= MacOS.pkgutil_info("com.dortania.opencore-legacy-patcher").present?
-          return unless has_opencore
+          return if ::Hardware::CPU.physical_cpu_arm64?
+          return if ENV["CI"]
+
+          # https://dortania.github.io/OpenCore-Legacy-Patcher/UPDATE.html#checking-oclp-and-opencore-versions
+          begin
+            opencore_version = Utils.safe_popen_read("/usr/sbin/nvram",
+                                                     "4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102:opencore-version")
```
Comment (EricFromCanada) [inline Library/Homebrew/extend/os/mac/diagnostic.rb:138]:
```suggestion
                                                     "4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102:opencore-version").split[1]
```
Save just the version string.

**Comment by MikeMcQuaid** [comment]
> If that's the case then maybe we can allow it in CI?

Yeh, agreed to allow in CI unless it's failing specifically in our CI.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good once CI check removed.

File: Library/Homebrew/extend/os/mac/diagnostic.rb [hunk 2]
```diff
@@ -129,11 +129,17 @@ def check_for_unsupported_macos
...
+            return if opencore_version.blank?
+          rescue ErrorDuringExecution
+            return
+          end
 
           <<~EOS
             You have installed macOS using OpenCore Legacy Patcher.
```
Comment (EricFromCanada) [inline Library/Homebrew/extend/os/mac/diagnostic.rb:145]:
```suggestion
            You have booted macOS using OpenCore Legacy Patcher.
```

**Review by EricFromCanada — APPROVED** [review]
I can confirm that `nvram 4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102:opencore-version` only returns a value when the current installation was actually booted via OCLP. (Oddly, the other command `nvram 4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102:OCLP-Version` still returns a value even when booting off of a non-OCLP boot disk.)

An alternate detection method would be to [check for Lilu in the list of loaded kexts](https://dortania.github.io/OpenCore-Legacy-Patcher/PATCHEXPLAIN.html#injected-kexts), e.g.:
```
% kmutil showloaded | grep Lilu
No variant specified, falling back to release
   52   10 0                  0x2f000    0x2f000    as.vit9696.Lilu (1.6.7) 087C34D2-49F8-3FDA-8757-27E9425C9EE4 <9 7 6 3 2 1>
```

**Comment by ZhongRuoyu** [comment]
> Looks good once CI check removed.

@MikeMcQuaid Not sure if I followed. I added the CI check (`ENV["CI"]`) to allow CI to pass because our self-hosted runners do seem to be on OpenCore. Would you like it to be done differently?

**Comment by MikeMcQuaid** [comment]
> Not sure if I followed. I added the CI check (`ENV["CI"]`) to allow CI to pass because our self-hosted runners do seem to be on OpenCore. Would you like it to be done differently?

Ah, no worries, ignore me then. 🚢 it!

**Comment by Bo98** [comment]
> Interesting that CI runner triggers it. I wonder if it is actually set up with OpenCore.

OpenCore is a bootloader whereas OCLP is what patches old hardware to run newer OS.
We use a custom bootloader because we run multiple instances of macOS on the same hardware.

So the unsupported configuration is more OCLP rather than OpenCore as that is what makes old CPUs run x86_64 binaries that might not necessarily be built for said old CPUs (i.e. pre-Westmere). Given Eric's findings above, the reliable check might be to check if _both_ `OCLP-Version` and `opencore-version` is set.

**Comment by ZhongRuoyu** [comment]
> OpenCore is a bootloader whereas OCLP is what patches old hardware to run newer OS. We use a custom bootloader in CI because we run multiple instances of macOS under a k8s environment on native hardware (so the host layer is technically Linux). We do not use OCLP.
> 
> So the unsupported configuration is more OCLP rather than OpenCore as that is what makes old CPUs run on macOS versions where our x86_64 binaries might not necessarily be built for said old CPUs (i.e. pre-Westmere). Given Eric's findings above, the reliable check might be to check if _both_ `OCLP-Version` and `opencore-version` is set.

Thanks for the explanation! Adding an additional check for OCLP in #18766. I suppose with that we won't need to skip the check in CI.

**Pull request merged**